src/kv/test/kv_dynamic_tables.cpp

```diff
@@ -0,0 +1,501 @@
+    REQUIRE(foo_v.has_value());
+    REQUIRE(foo_v.value() == "foo");
+
+    const auto bar_v = view->get("bar");
+    REQUIRE(bar_v.has_value());
+    REQUIRE(bar_v.value() == "bar");
+  }
+}
+
+TEST_CASE("Mid rollback safety" * doctest::test_suite("dynamic"))
+{
+  kv::Store kv_store;
+
+  auto encryptor = std::make_shared<kv::NullTxEncryptor>();
+  kv_store.set_encryptor(encryptor);
+
+  constexpr auto map_name = "my_new_map";
+
+  const auto version_before = kv_store.current_version();
+
+  {
+    auto tx = kv_store.create_tx();
+
+    auto view = tx.get_view2<MapTypes::StringString>(map_name);
+    view->put("foo", "bar");
+
+    REQUIRE(tx.commit() == kv::CommitSuccess::OK);
+  }
+
+  {
+    auto tx = kv_store.create_tx();
+    auto view = tx.get_view2<MapTypes::StringString>(map_name);
+    const auto v_0 = view->get("foo");
+    REQUIRE(v_0.has_value());
+    REQUIRE(v_0.value() == "bar");
+
+    // Rollbacks may happen while a tx is executing, and these can delete the
+    // maps this tx is executing over
+    kv_store.rollback(version_before);
+
+    const auto v_1 = view->get("foo");
+    REQUIRE(v_0.has_value());
+    REQUIRE(v_0.value() == "bar");
+
+    auto view_after = tx.get_view2<MapTypes::StringString>(map_name);
+    REQUIRE(view_after == view);
+
+    view->put("foo", "baz");
+
+    const auto result = tx.commit();
+    REQUIRE(result == kv::CommitSuccess::CONFLICT);
+  }
+}
+
+TEST_CASE(
+  "Security domain is determined by map name" * doctest::test_suite("dynamic"))
+{
+  kv::Store kv_store;
+
+  auto encryptor = std::make_shared<kv::NullTxEncryptor>();
+  kv_store.set_encryptor(encryptor);
+
+  {
+    auto tx = kv_store.create_tx();
+    auto view = tx.get_view2<MapTypes::StringString>("public:foo");
+    view->put("foo", "bar");
+
+    REQUIRE(tx.commit() == kv::CommitSuccess::OK);
+  }
+
+  {
+    auto tx = kv_store.create_tx();
+    auto view = tx.get_view2<MapTypes::StringString>("foo");
+    view->put("hello", "world");
+
+    REQUIRE(tx.commit() == kv::CommitSuccess::OK);
+  }
+
+  {
+    auto public_map = kv_store.get<kv::untyped::Map>("public:foo");
+    REQUIRE(public_map != nullptr);
+    REQUIRE(public_map->get_security_domain() == kv::SecurityDomain::PUBLIC);
+
+    auto private_map = kv_store.get<kv::untyped::Map>("foo");
+    REQUIRE(private_map != nullptr);
+    REQUIRE(private_map->get_security_domain() == kv::SecurityDomain::PRIVATE);
+  }
+
+  {
+    auto tx = kv_store.create_tx();
+    auto [public_view, private_view] =
+      tx.get_view2<MapTypes::StringString, MapTypes::StringString>(
+        "public:foo", "foo");
+
+    // These are _different views_ over _different maps_
+    REQUIRE(public_view != private_view);
+
+    const auto pub_v = public_view->get("foo");
+    REQUIRE(pub_v.has_value());
+    REQUIRE(pub_v.value() == "bar");
+
+    const auto priv_v = private_view->get("hello");
+    REQUIRE(priv_v.has_value());
+    REQUIRE(priv_v.value() == "world");
+  }
```

Comment:
I may be useful to additionally check that dynamic maps happily survive being swapped from one store to another (which happens during recovery, in the public -> private phase).